solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,420 @@
+        );
+
+        OptimisticMintingRequest storage request = optimisticMintingRequests[
+            depositKey
+        ];
+        require(
+            request.requestedAt != 0,
+            "Optimistic minting not requested for the deposit"
+        );
+        require(
+            request.finalizedAt == 0,
+            "Optimistic minting already finalized for the deposit"
+        );
+
+        // Delete it. It allows to request optimistic minting for the given
+        // deposit again. Useful in case of an errant Guarian.
+        delete optimisticMintingRequests[depositKey];
+
+        emit OptimisticMintingCancelled(
+            msg.sender,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    /// @notice Adds the address to the Minter set.
+    function addMinter(address minter) external onlyOwner {
+        require(!isMinter[minter], "This address is already a minter");
+        isMinter[minter] = true;
+        emit MinterAdded(minter);
+    }
+
+    /// @notice Removes the address from the Minter set.
+    function removeMinter(address minter) external onlyOwnerOrGuardian {
+        require(isMinter[minter], "This address is not a minter");
+        delete isMinter[minter];
+        emit MinterRemoved(minter);
+    }
+
+    /// @notice Adds the address to the Guardian set.
+    function addGuardian(address guardian) external onlyOwner {
+        require(!isGuardian[guardian], "This address is already a guardian");
+        isGuardian[guardian] = true;
+        emit GuardianAdded(guardian);
+    }
+
+    /// @notice Removes the address from the Guardian set.
+    function removeGuardian(address guardian) external onlyOwner {
+        require(isGuardian[guardian], "This address is not a guardian");
+        delete isGuardian[guardian];
+        emit GuardianRemoved(guardian);
+    }
+
+    /// @notice Pauses the optimistic minting. Note that the pause of the
+    ///         optimistic minting does not stop the standard minting flow
+    ///         where wallets sweep deposits.
+    function pauseOptimisticMinting() external onlyOwner {
+        require(
+            !isOptimisticMintingPaused,
+            "Optimistic minting already paused"
+        );
+        isOptimisticMintingPaused = true;
+        emit OptimisticMintingPaused();
+    }
+
+    /// @notice Unpauses the optimistic minting.
+    function unpauseOptimisticMinting() external onlyOwner {
+        require(isOptimisticMintingPaused, "Optimistic minting is not paused");
+        isOptimisticMintingPaused = false;
+        emit OptimisticMintingUnpaused();
+    }
+
+    /// @notice Calculates deposit key the same way as the Bridge contract.
+    ///         The deposit key is computed as
+    ///         keccak256(fundingTxHash | fundingOutputIndex).
+    function calculateDepositKey(
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex
+    ) public view returns (uint256) {
```

Comment:
> Shouldn't this be a pure function?

Yes, done!

> Apart from this could this function be abstracted to the Deposit library, so deposit keys are always computed this way? 

This is true but I'd prefer to do it separately. This PR is already huge.